docs/proposals/set-ref.md

```diff
@@ -0,0 +1,143 @@
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# ControllerRef proposal
+
+Author: gmarek@
+Last edit: 2016-05-06
+Status: raw
+
+Approvers:
+- [ ] briangrant
+- [ ] dbsmith
+
+**Table of Contents**
+
+- [Goal of CollectionReference](#goal-of-setreference)
+- [Non goals](#non-goals)
+- [API and semantic changes](#api-and-semantic-changes)
+- [Upgrade/downgrade procedure](#upgradedowngrade-procedure)
+- [Orphaning/adoption](#orphaningadoption)
+- [Implementation plan (sketch)](#implementation-plan-sketch)
+- [Considered alternatives](#considered-alternatives)
+
+# Goal of CollectionReference
+
+Main goal of *CollectionReference* effort is to solve a problem of overlapping controllers that fight over some resources (e.g. *ReplicaSets* fighting with *ReplicationControllers* over *Pods*), which cause serious [problems](https://github.com/kubernetes/kubernetes/issues/24433) such as exploding memory of Controller Manager.
+
+We don’t want to have (just) an in-memory solution, as we don’t want a Controller Manager crash to cause massive changes in object ownership in the system. I.e. we need to persist the information about "owning controller".
+
+Secondary goal of this effort is to improve performance of various controllers and schedulers, by removing the need for expensive lookup for all matching "controllers".
+
+# Non goals
+
+Cascading deletion is not a goal of this effort. Cascading deletion will use *ownerReferences*, which is a [separate effort](garbage-collection.md).
+
+There will be some interactions between controller and parent references, they will share a common orphaning/adoption library and *CollectionRef* may be treated as one of *OwnerReferencess*.
+
+# API and semantic changes
+
+There will be a new API field in the *ObjectMeta* in which we will store a *CollectionReference*:
+
+```
+ObjectMeta {
+    …
+    CollectionRef *CollectionReference
+    …
+}
+```
+
+```
+type CollectionReference struct {
+    // Version of the referent in group/version format.
+    APIVersion string
+    // Kind of the referent.
+    Kind string
+    // Name of the referent.
+    Name string
+    // UID of the referent.
+    UID types.UID
+    // Should this CollectionReference be treated as an owner for cascading deletion purposes
+    Owner bool
```

Comment:
I think if you want the garbage collector to be responsible for clearing these, then we should really consider making these less different from the things the garbage collector actually looks at.